public/src/components/epicTests/epicTestsForm.tsx

```
@@ -138,26 +165,83 @@ class EpicTestsForm extends React.Component<EpicTestFormProps, EpicTestsFormStat
+        return (
+          <>
+            <Typography className={classes.warning}>File locked for editing by {friendlyName} (<a href={"mailto:" + this.state.lockStatus.email}>{this.state.lockStatus.email}</a>) at {friendlyTimestamp}</Typography>
+            <div className={classes.buttons}>
+              <ButtonWithConfirmationPopup
+                buttonText="Take control"
+                confirmationText={`Are you sure? Please tell ${friendlyName} that their unpublished changes will be lost.`}
+                onConfirm={this.requestTakeControl}
+                icon={<LockOpenIcon />}
+              />
+            </div>
+          </>
+        );
+      } else {
+        return (
+          <Button
+            onClick={this.requestEpicTestsLock}
+            variant="contained" color="primary"
+          >Edit</Button>
+        )
+      }
+    } else {
+      return (
+        <div>
+          <ButtonWithConfirmationPopup
+          buttonText="Publish"
+          confirmationText="Are you sure? This will replace all live tests!"
+          onConfirm={this.save}
+          icon={<CloudUploadIcon />}
+        />
+        <ButtonWithConfirmationPopup
+        buttonText="Cancel"
+        confirmationText="Are you sure? All unpublished data will be lost!"
+        onConfirm={() => this.cancel()}
+        icon={<RefreshIcon />}
+        />
+        </div>
+        )
+    }
+  }
+
+  makeFriendlyDate = (timestamp: string): string => {
+    const datetime = new Date(timestamp);
+    return `${datetime.getHours()}:${datetime.getMinutes()} on ${datetime.getDate()}/${datetime.getMonth() + 1}/${datetime.getFullYear()}`;
+  }
+
+  makeFriendlyName = (email: string): string | undefined => {
+    const nameArr: RegExpMatchArray | null = email.match(/^([a-z]*)\.([a-z]*).*@.*/);
+    return nameArr ? `${nameArr[1][0].toUpperCase()}${nameArr[1].slice(1,)} ${nameArr[2][0].toUpperCase()}${nameArr[2].slice(1,)}` : undefined;
```

how about instead of returning `undefined` if the match fails, just return `email` as is?

---

The UI would then display the email twice in that case - but I guess that won't really happen in practice.

---

oh right, I didn't realise it's already displaying the email!

---

👍 looks good!

---

Only on the main page (with link), but displays the name alone in the popup warning message.